### PR TITLE
Fix saving of bool and null values to the database

### PR DESCRIPTION
**Pull request by koca2000**
The `Database::val` method did not work correctly with bool values. They were interpreted as strings "1" and "" which is not valid for database. It is now converted to integer instead of string.

The `Database::changesetMapToList` method did not work properly with null values as they were converted to an empty string when used as key in an associative array. It now uses the `Database::val` and `Database::raw` methods to avoid this problem.

### CODE REVIEW

File: system/class/Database/Database.php [hunk 1]
```diff
@@ -606,7 +610,7 @@ static function changesetMapToList(array $changesetMap): array
 
         foreach ($ids as $id) {
             foreach ($changesetMap[$id] as $column => $value) {
-                $commonChanges[$column][$value][$id] = true;
+                $commonChanges[$column][self::val($value)][$id] = true;
```
Comment (Shira-3749) [inline system/class/Database/Database.php:613]: This should not be needed as boolean keys already get converted to integers. Then the `self::raw()` calls also become unnecessary.

File: system/class/Database/Database.php [hunk 2]
```diff
@@ -352,6 +352,10 @@ static function val($value): string
             return $value->getSql();
         }
 
+        if (is_bool($value)) {
+            $value = (int)$value;
```
Comment (Shira-3749) [inline system/class/Database/Database.php:356]: At this point the final value should just be returned, there's no need for it to get processed further in the `is_numeric()` case below.

```suggestion
            return (int) $value;
```

**Review by Shira-3749 — CHANGES_REQUESTED** [review]

**Comment by koca2000** [comment]
The use of `Database::val` and `Database::raw` methods is still necessary for both bool and null values because without it the value would be converted to string ("" in case of null and "1" or "" in case of bool) which would be later interpreted as string in subsequent `Database::val` call.

File: system/class/Database/Database.php [hunk 3]
```diff
@@ -352,6 +352,10 @@ static function val($value): string
             return $value->getSql();
         }
 
+        if (is_bool($value)) {
+            return (int) $value;
```
Comment (Shira-3749) [inline system/class/Database/Database.php:356]: My bad, there is a `string` return type.

```suggestion
            return $value ? '1' : '0';
```

**Review by Shira-3749 — APPROVED** [review]

**Pull request merged**